### PR TITLE
Add APPROX_HISTOGRAM_K Operation

### PR DESCRIPTION
**Pull request by jbrooks-stripe**
## Summary

Adds an APPROX_HISTOGRAM_K operation based on the FrequentItems Sketch:
https://datasketches.apache.org/docs/Frequency/FrequentItemsOverview.html

This is mostly a wrapper on top of the frequent items sketch, but have made the two following modifications to make it somewhat more user friendly:
* The sketch requires a k which is a power of two. Instead of making this a user requirement, we will take the next closest power of two and truncate down to k in the results.
* The underlying map used by the sketch has a load factor of 0.75, which causes the approximation to kick in before there are k elements. We changed this to keep an exact histogram/map and then switch over to the sketch once the entries exceed k.

## Why / Goal

The current histogram operation uses unbounded memory and isn't stable for production use cases. 

## Test Plan

We've end to end tested everything through backfills/CLI on the batch side on our end.

- [x] Added Unit Tests
- [x] Covered by existing CI
- [x] Integration tested

## Checklist
- [x] Documentation update

### CODE REVIEW

File: aggregator/src/main/scala/ai/chronon/aggregator/base/SimpleAggregators.scala [hunk 1]
```diff
@@ -1,35 +1,14 @@
-/*
```
Comment (nikhilsimha) [inline aggregator/src/main/scala/ai/chronon/aggregator/base/SimpleAggregators.scala:1]: Is this intentional?

File: aggregator/src/main/scala/ai/chronon/aggregator/row/ColumnAggregator.scala [hunk 2]
```diff
@@ -375,6 +377,17 @@ object ColumnAggregator {
       case Operation.LAST    => timed(new Last(inputType))
       case Operation.FIRST_K => timed(new FirstK(inputType, aggregationPart.getInt("k")))
       case Operation.LAST_K  => timed(new LastK(inputType, aggregationPart.getInt("k")))
+      case Operation.APPROX_HISTOGRAM_K =>
```
Comment (nikhilsimha) [inline aggregator/src/main/scala/ai/chronon/aggregator/row/ColumnAggregator.scala:380]: Can you move this up closer to histogram?

File: aggregator/src/main/scala/ai/chronon/aggregator/base/SimpleAggregators.scala [hunk 3]
```diff
@@ -459,8 +465,7 @@ class ApproxDistinctCount[Input: CpcFriendly](lgK: Int = 8) extends SimpleAggreg
     CpcSketch.heapify(normalized.asInstanceOf[Array[Byte]])
 }
 
-class ApproxPercentiles(k: Int = 128, percentiles: Array[Double] = Array(0.5))
-    extends SimpleAggregator[Float, KllFloatsSketch, Array[Float]] {
+class ApproxPercentiles(k: Int = 128, percentiles: Array[Double] = Array(0.5)) extends SimpleAggregator[Float, KllFloatsSketch, Array[Float]] {
```
Comment (nikhilsimha) [inline aggregator/src/main/scala/ai/chronon/aggregator/base/SimpleAggregators.scala:468]: Intentional?

File: spark/src/main/scala/ai/chronon/spark/ChrononKryoRegistrator.scala [hunk 4]
```diff
@@ -1,28 +1,15 @@
-/*
```
Comment (nikhilsimha) [inline spark/src/main/scala/ai/chronon/spark/ChrononKryoRegistrator.scala:1]: Intentional?

**Review by nikhilsimha — COMMENTED** [review]
Few comments. But the approach looks sound overall.

File: aggregator/src/main/scala/ai/chronon/aggregator/base/SimpleAggregators.scala (hunk 1 above)
Comment (jbrooks-stripe) [reply aggregator/src/main/scala/ai/chronon/aggregator/base/SimpleAggregators.scala]: Not intentional. Updating merge conflicts. Will tag you when everything is ready.

File: aggregator/src/main/scala/ai/chronon/aggregator/base/SimpleAggregators.scala [hunk 5]
```diff
@@ -372,7 +373,12 @@ class FrequentItems[T: FrequentItemsFriendly](val mapSize: Int, val errorType: E
   override def irType: DataType = BinaryType
   type Sketch = ItemsSketch[T]
   override def prepare(input: T): Sketch = {
-    val sketch = new ItemsSketch[T](mapSize)
+    // The ItemsSketch implementation requires a size with a positive power of 2
+    // Initialize the sketch with the next closest power of 2
+    val power = math.ceil(math.log(math.max(mapSize, 1)) / math.log(2))
```
Comment (nikhilsimha) [inline aggregator/src/main/scala/ai/chronon/aggregator/base/SimpleAggregators.scala:378]: ChatGpt gave me 

int nextPowerOf2 = (n > 1) ? Integer.highestOneBit(n - 1) << 1 : 1;

Seems more readable I think.

File: spark/src/main/scala/ai/chronon/spark/ChrononKryoRegistrator.scala [hunk 6]
```diff
@@ -127,7 +126,6 @@ class ChrononKryoRegistrator extends KryoRegistrator {
       "scala.reflect.ManifestFactory$LongManifest",
       "org.apache.spark.sql.execution.joins.EmptyHashedRelation$",
       "scala.reflect.ManifestFactory$$anon$1",
-      "scala.reflect.ClassTag$GenericClassTag",
```
Comment (nikhilsimha) [inline spark/src/main/scala/ai/chronon/spark/ChrononKryoRegistrator.scala:130]: This might cause issues based on version of scala we are testing with. We recently disabled scala 2.11 tests - which is probably the version that removing this will break.

Let's remove this line.

File: aggregator/src/main/scala/ai/chronon/aggregator/base/SimpleAggregators.scala (hunk 5 above)
Comment (nikhilsimha) [inline aggregator/src/main/scala/ai/chronon/aggregator/base/SimpleAggregators.scala:378]: It might be worth adding tests to this line at various boundary conditions - -1, 0, 1, 31, 32, 33

**Review by nikhilsimha — COMMENTED** [review]
Let's also add a test to spark/test/FetcherTest.scala 

Looks good otherwise. Mostly non-blocking comments.

**Comment by jbrooks-stripe** [comment]
> Let's also add a test to spark/test/FetcherTest.scala
> 
> Looks good otherwise. Mostly non-blocking comments.

Appreciate you taking a look. Will address your comments and some from our internal repo shortly.

**Review by andrewlee-stripe — APPROVED** [review]
